state-chain/pallets/cf-cfe-event-emitter/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod benchmarking;
+mod weights;
+
+pub use weights::PalletWeight;
+use weights::WeightInfo;
+
+use cf_chains::{
+	btc::BitcoinCrypto, dot::PolkadotCrypto, evm::EvmCrypto, Bitcoin, Ethereum, Polkadot,
+};
+use cf_primitives::{Ed25519PublicKey, Ipv6Addr, Port};
+use cf_traits::{CfeEventEmitterForChain, CfeEventEmitterForCrypto, CfeEventEmitterT, Chainflip};
+use frame_support::{
+	pallet_prelude::Hooks, storage::types::StorageMap, traits::StorageVersion, Twox64Concat,
+};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_std::vec::Vec;
+
+pub const PALLET_VERSION: StorageVersion = StorageVersion::new(1);
+
+/// How long to keep cfe events for (in SC blocks)
+const EVENT_LIFETIME: u32 = 20;
+
+pub type EventId = u64;
+
+pub type CfeEvent<T> = cfe_events::CfeEvent<<T as Chainflip>::ValidatorId>;
+pub type ThresholdSignatureRequest<T, C> =
+	cfe_events::ThresholdSignatureRequest<<T as Chainflip>::ValidatorId, C>;
+pub type KeyHandoverRequest<T, C> =
+	cfe_events::KeyHandoverRequest<<T as Chainflip>::ValidatorId, C>;
+pub type KeygenRequest<T> = cfe_events::KeygenRequest<<T as Chainflip>::ValidatorId>;
+pub type TxBroadcastRequest<T, C> =
+	cfe_events::TxBroadcastRequest<<T as Chainflip>::ValidatorId, C>;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use cf_traits::Chainflip;
+
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::storage_version(PALLET_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: Chainflip {
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_cfe_events)]
+	#[pallet::unbounded]
+	pub type CfeEvents<T: Config> =
+		StorageMap<_, Twox64Concat, BlockNumberFor<T>, Vec<CfeEvent<T>>>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(block_number: BlockNumberFor<T>) -> frame_support::weights::Weight {
+			use frame_support::sp_runtime::Saturating;
+
+			CfeEvents::<T>::remove(block_number.saturating_sub(EVENT_LIFETIME.into()));
```

Comment:
Related to above, this won't be necessary as we'd just overwrite the existing the events in the storagevalue - rather than have a storage map